pkg/apis/rollouts/validation/validation_references_test.go

```diff
@@ -0,0 +1,271 @@
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/ghodss/yaml"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+)
+
+const successCaseVsvc = `apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: istio-vsvc
+  namespace: default
+spec:
+  gateways:
+  - istio-rollout-gateway
+  hosts:
+  - istio-rollout.dev.argoproj.io
+  http:
+  - name: primary
+    route:
+    - destination:
+        host: 'stable'
+      weight: 100
+    - destination:
+        host: canary
+	  weight: 0
+  - name: secondary
+	route:
+	- destination:
+		host: 'stable'
+	  weight: 100
+	- destination:
+	    host: canary
+	  weight: 0`
+
+const failCaseVsvc = `apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: istio-vsvc
+  namespace: default
+spec:
+  gateways:
+  - istio-rollout-gateway
+  hosts:
+  - istio-rollout.dev.argoproj.io
+  http:
+  - name: primary
+    route:
+    - destination:
+        host: 'not-stable'
+      weight: 100
+    - destination:
+        host: canary
+	  weight: 0
+  - name: secondary
+	route:
+	- destination:
+		host: 'not-stable'
+	  weight: 100
+	- destination:
+	    host: canary
+	  weight: 0`
+
+var analysisTemplate = v1alpha1.AnalysisTemplate{
+	ObjectMeta: metav1.ObjectMeta{
+		Name: "analysis-template-name",
+	},
+	Spec: v1alpha1.AnalysisTemplateSpec{
+		Metrics: []v1alpha1.Metric{{
+			Name:     "metric-name",
+			Interval: "1m",
+			Count:    1,
+		}},
+	},
+}
+
+var analysisTemplateWithType = AnalysisTemplateWithType{
+	AnalysisTemplate: &analysisTemplate,
+	TemplateType:     CanaryStep,
+	AnalysisIndex:    0,
+	CanaryStepIndex:  0,
+}
+
+var ro = &v1alpha1.Rollout{
+	Spec: v1alpha1.RolloutSpec{
+		Strategy: v1alpha1.RolloutStrategy{
+			Canary: &v1alpha1.CanaryStrategy{
+				StableService: "stable-service-name",
+				TrafficRouting: &v1alpha1.RolloutTrafficRouting{
+					ALB: &v1alpha1.ALBTrafficRouting{
+						Ingress: "alb-ingress",
+					},
+				},
+			},
+		},
+	},
+}
+
+var ingress = v1beta1.Ingress{
+	ObjectMeta: metav1.ObjectMeta{
+		Name: "alb-ingress",
+	},
+	Spec: v1beta1.IngressSpec{
+		Rules: []v1beta1.IngressRule{{
+			Host: "fakehost.example.com",
+			IngressRuleValue: v1beta1.IngressRuleValue{
+				HTTP: &v1beta1.HTTPIngressRuleValue{
+					Paths: []v1beta1.HTTPIngressPath{{
+						Path: "/foo",
+						Backend: v1beta1.IngressBackend{
+							ServiceName: "stable-service-name",
+							ServicePort: intstr.FromString("use-annotations"),
+						},
+					}},
+				},
+			},
+		}},
+	},
+}
+
+var svc = ServiceWithType{
+	Service: &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "stable-service-name",
+		},
+	},
+	Type: StableService,
+}
+
+func strToUnstructured(yamlStr string) *unstructured.Unstructured {
```

Comment:
You should be able to use https://github.com/argoproj/argo-rollouts/blob/master/utils/unstructured/unstructured.go#L8 instead of reimplementing this method.